Flight Analysis

Profs notes on the draft: Excellent plan. It will be great to report/show pre-COVID vs. after-COVID analysis (and... even during the COVID in each month/quarterly). You can also try both linear and non-linear regression models (e.g., linear regression and SVR regression with nonlinear kernel). https://scikit-learn.org/stable/auto_examples/svm/plot_svm_regression.html

The data: 


**All imports needed to run**

If the next snippet fails, go into Runtime>> Restars Runtime

In [1]:
#Only Needs to be run once
# import sys
# !{sys.executable} -m pip install pymongo
# !{sys.executable} -m pip install dnspython==2.1.0
# !{sys.executable} -m pip install scikit-learn


### Imports 

In [2]:
import pymongo
import dns


import numpy as np
import pandas as pd
import json

from sklearn import linear_model
from sklearn.linear_model import LinearRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVR

from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.metrics import mean_squared_error, r2_score, accuracy_score


import matplotlib.pyplot as plt

import joblib
import pickle

**Importing Data into MongoDB**

We downloaded the data from the website <<INSERT HERE>>

We tried loading it through Drive/Google Drive to MongoDB. However, our notebook was not able to handle the amount of data, because it ran out of space.

So we used, MongoDB Compass to upload the existing CSVs to our collection
    
We made a duplicate of the original datatset, because our work requires deleting some values

**Connecting to MongoDB**

In [3]:
# RUN
client = pymongo.MongoClient("mongodb+srv://dbUser:test123@cluster0.92wx0.mongodb.net/finalProject?retryWrites=true&w=majority")

db = client['finalProject']

collection = db['flightData']

# pipeline = [ {"$match": {}}, {"$out": "anotheDuplicate"},]
# collection.aggregate(pipeline)

**DELETING DATA**
Deleting the data, which has no values for any of the analysis factors needed below

In [4]:
#Only Needs to be run once
# collection.delete_many({'MONTH' : '6' } )
# collection.delete_many({'MONTH' : '7' } )
# collection.delete_many({'DEP_DELAY' : {'$eq': None}} )
# collection.delete_many({'DEP_DELAY_GROUP' : {'$eq': None}} )
# collection.delete_many({'ARR_DELAY_GROUP' : {'$eq': None}} )
# collection.delete_many({'YEAR' : {'$eq': None}} )
# collection.delete_many({'ORIGIN_AIRPORT_ID' : {'$eq': None}} )
# collection.delete_many({'DEST_AIRPORT_ID' : {'$eq': None}} )
# collection.delete_many({'DEP_TIME' : {'$eq': None}} )
# collection.delete_many({'ARR_TIME' : {'$eq': None}} )
# collection.delete_many({'ORIGIN' : {'$eq': None}} )

In [ ]:
print(collection.count_documents({}))

**Looking at Data**

In [5]:
#Just Checking
allPreCovid = collection.find({'YEAR':'2019'})
allCovid = collection.find({'YEAR':'2020'})

print("Number of data for Pre-Covid: June, July, August of 2019")
#print(allPreCovid.count())
print(collection.count_documents({'YEAR':'2019'}))

print()
print("Number of data for Covid: June, July, August of 2020")
#print(allCovid.count())
print(collection.count_documents({'YEAR':'2020'}))

print()
print("Example of one data point")
print(collection.find_one())

Number of data for Pre-Covid: June, July, August of 2019
645351

Number of data for Covid: June, July, August of 2020
371969

Example of one data point
{'_id': ObjectId('608de2ba9c0bff4d24c42081'), 'YEAR': '2019', 'QUARTER': '3', 'MONTH': '8', 'DAY_OF_MONTH': '25', 'DAY_OF_WEEK': '7', 'FL_DATE': '2019-08-25', 'OP_UNIQUE_CARRIER': 'WN', 'TAIL_NUM': 'N433LV', 'OP_CARRIER_FL_NUM': '4349', 'ORIGIN_AIRPORT_ID': '11292', 'ORIGIN_AIRPORT_SEQ_ID': '1129202', 'ORIGIN_CITY_MARKET_ID': '30325', 'ORIGIN': 'DEN', 'ORIGIN_CITY_NAME': 'Denver, CO', 'ORIGIN_STATE_ABR': 'CO', 'ORIGIN_STATE_NM': 'Colorado', 'DEST_AIRPORT_ID': '13232', 'DEST_AIRPORT_SEQ_ID': '1323202', 'DEST_CITY_MARKET_ID': '30977', 'DEST': 'MDW', 'DEST_CITY_NAME': 'Chicago, IL', 'DEST_STATE_ABR': 'IL', 'DEST_STATE_NM': 'Illinois', 'CRS_DEP_TIME': '2150', 'DEP_TIME': '2150', 'DEP_DELAY': '0.00', 'DEP_DELAY_NEW': '0.00', 'DEP_DEL15': '0.00', 'DEP_DELAY_GROUP': '0', 'DEP_TIME_BLK': '2100-2159', 'TAXI_OUT': '12.00', 'WHEELS_OFF': '2202', '

**Analyzing**

### Gaussian Naive Bayes

In [6]:
def encoding(year, enc):
    origin = collection.find({'YEAR':year},{'_id':0,enc:1})
    ids = []
    for doc in origin:
        ids+=[(doc.get(enc))]
    return ids

In [7]:
#ENCODINGS 2019

leOrigin = preprocessing.LabelEncoder()
ids = encoding('2019', 'ORIGIN')
encOrigin19 =leOrigin.fit_transform(ids)

le_dictOrigin19 = dict(zip(leOrigin.classes_, leOrigin.transform(leOrigin.classes_)))

leDeptTime = preprocessing.LabelEncoder()
ids = encoding('2019', 'DEP_TIME')
encDeptTime19=leDeptTime.fit_transform(ids)
le_dictDeptTime19 = dict(zip(leDeptTime.classes_, leDeptTime.transform(leDeptTime.classes_)))

In [8]:
#ENCODINGS 2020

leOrigin = preprocessing.LabelEncoder()
ids = encoding('2020', 'ORIGIN')
encOrigin20 =leOrigin.fit_transform(ids)

le_dictOrigin20 = dict(zip(leOrigin.classes_, leOrigin.transform(leOrigin.classes_)))

leDeptTime = preprocessing.LabelEncoder()
ids = encoding('2020', 'DEP_TIME')
encDeptTime20=leDeptTime.fit_transform(ids)
le_dictDeptTime20 = dict(zip(leDeptTime.classes_, leDeptTime.transform(leDeptTime.classes_)))

In [9]:
X19 = collection.find({'YEAR':'2019'},{'_id':0})
Y19 = collection.find({'YEAR':'2019'},{'_id':0,'DEP_DELAY_GROUP':1})
# print(Y19.count())
# print(X19.count())

In [10]:
X20 = collection.find({'YEAR':'2020'},{'_id':0})
Y20 = collection.find({'YEAR':'2020'},{'_id':0,'DEP_DELAY_GROUP':1})
# print(Y20.count())
# print(X20.count())

In [11]:
ids = []
for doc in Y19:
    ids+=[float(doc.get('DEP_DELAY_GROUP'))]

labels19 = np.array(ids)

i=0
xs = []
for doc in X19:
    xs.append([encOrigin19.item(i), encDeptTime19.item(i)])
    i+=1

features19 = np.array(xs)

x_train, x_test, y_train, y_test = train_test_split(features19, labels19, test_size=0.5, random_state=109)

gnb19 = GaussianNB()
gnb19.fit(x_train, y_train)

y_pred = gnb19.predict(x_test)

print("2019 Number of mislabeled points out of a total %d points : %d"
      % (x_test.shape[0], (y_test != y_pred).sum()))

print("2019 Accuracy:",metrics.accuracy_score(y_test, y_pred))

2019 Number of mislabeled points out of a total 322676 points : 134531
2019 Accuracy: 0.5830771423967075


In [ ]:
ids = []
for doc in Y20:
    ids+=[float(doc.get('DEP_DELAY_GROUP'))]

labels20 = np.array(ids)

i=0
xs = []
for doc in X20:
    xs.append([encOrigin20.item(i), encDeptTime20.item(i)])
    i+=1

features20 = np.array(xs)

x_train, x_test, y_train, y_test = train_test_split(features20, labels20, test_size=0.5, random_state=109)

gnb20 = GaussianNB()
gnb20.fit(x_train, y_train)

y_pred = gnb20.predict(x_test)

print("2020 Number of mislabeled points out of a total %d points : %d"
      % (x_test.shape[0], (y_test != y_pred).sum()))

print("2020 Accuracy:",metrics.accuracy_score(y_test, y_pred))

In [ ]:
def predictFlight(origin, dest, depTime):
    originAirport = le_dictOrigin19.get(origin, '<Unknown>')
    #destAirport = le_dict
    departureTime = le_dictDeptTime19.get(depTime, '')
    features19 = np.array([[originAirport,departureTime ]])
    
    originAirport = le_dictOrigin20.get(origin, '<Unknown>')
    #destAirport = le_dict
    departureTime = le_dictDeptTime20.get(depTime, '')
    features20 = np.array([[originAirport,departureTime ]])
    print('Pre-Covid')
    print(gnb19.predict(features19))
    
    print()
    print('Covid')
    print(gnb20.predict(features20))

In [ ]:
predictFlight('BOS', 'DEN', '2020')

*Linear Regression*

X/Explanatory Variable

*   Origin City



Y/Dependent Variable


*   Delay Time


https://stackoverflow.com/questions/43288550/iopub-data-rate-exceeded-in-jupyter-notebook-when-viewing-image

In [ ]:
# Preparing data for Linear Regression
#ORIGIN_AIRPORT_ID
X = collection.find({'YEAR':'2019'},{'_id':0,'DEST_AIRPORT_ID':1})
Y = collection.find({'YEAR':'2019'},{'_id':0,'ARR_DELAY':1})

ids = []
for doc in Y:
    ids+=[float(doc.get('ARR_DELAY'))]

Y = np.array(ids)

xs = []
for doc in X:
    xs.append([int(doc.get('DEST_AIRPORT_ID'))]) #,int(doc.get( 'DEST_AIRPORT_ID'))])

X = np.array(xs)
x_train, x_test, y_train, y_test = train_test_split(X,Y, test_size=0.3)
regr = LinearRegression()
regr.fit(x_train, y_train)

y_pred = regr.predict(x_test)

# example from : https://scikit-learn.org/stable/auto_examples/linear_model/plot_ols.html

# The coefficients
print('Coefficients: \n', regr.coef_)
# The mean squared error
print('Mean squared error: %.2f'
      % mean_squared_error(y_test, y_pred))
# The coefficient of determination: 1 is perfect prediction
print('Coefficient of determination: %.2f'
      % r2_score(y_test, y_pred))

#print('Accuracy: ', accuracy_score(y_test, y_pred) )
# Plot outputs
plt.scatter(x_test, y_test)
plt.plot(x_test, y_pred, color='blue', linewidth=3)

plt.xticks(())
plt.yticks(())

### Multiple Linear Regression

In [ ]:
df = pd.DataFrame(list(collection.find({'YEAR':'2019', 'DAY_OF_MONTH': '25'})))

X = df[{'YEAR':'2019', 'DAY_OF_MONTH': '25'},{'_id':0,'ORIGIN_AIRPORT_ID':1, 'DEST_AIRPORT_ID':1}]
Y = df[{'YEAR':'2019', 'DAY_OF_MONTH': '25'},{'_id':0,'DEP_DELAY_GROUP':1}]
x, y = np.array(X), np.array(Y)


regr = linear_model.LinearRegression()
regr.fit(X, y)

predictedCO2 = regr.predict([['11292','13232']])

print(x)
print(y)

model = LinearRegression().fit(x, y)
r_sq = model.score(x, y)
print('coefficient of determination:', r_sq)
print('intercept:', model.intercept_)
print('slope:', model.coef_)

y_pred = model.predict(x)
print('predicted response:', y_pred, sep='\n')

y_pred = model.intercept_ + np.sum(model.coef_ * x, axis=1)


print('predicted response:', y_pred, sep='\n')

x_new = np.arange(10).reshape((-1, 2))
print(x_new)

y_new = model.predict(x_new)
print(y_new)


### SVR

In [12]:
#preCovidOC = collection.find({'YEAR':'2019'},{'_id':0,'ORIGIN_CITY_NAME':1})
#X = np.array(preCovidOC)
X = collection.find({'YEAR':'2019', 'DAY_OF_MONTH': '25'},{'_id':0,'ORIGIN_AIRPORT_ID':1, 'DEST_AIRPORT_ID':1})
#print(X)
Y = collection.find({'YEAR':'2019', 'DAY_OF_MONTH': '25'},{'_id':0,'DEP_DELAY_GROUP':1})
df = pd.DataFrame(list(collection.find({'YEAR':'2019', 'DAY_OF_MONTH': '25'})))

In [13]:
ids = []
for doc in Y:
    ids+=[float(doc.get('DEP_DELAY_GROUP'))]
Y = np.array(ids)
#print(Y)
#print(len(Y))
xs = []
for doc in X:
    xs.append([int(doc.get('ORIGIN_AIRPORT_ID'))])#,int(doc.get( 'DEST_AIRPORT_ID'))])
X = np.array(xs)
#print(X)

In [14]:
from sklearn.model_selection import train_test_split

x_train2, x_test2, y_train2, y_test2 = train_test_split(X,Y, test_size=0.3)

#### Linear SVR

In [15]:
svr_lin = SVR(kernel = 'linear')
svr_lin.fit(x_train2, y_train2)
y_predLin = svr_lin.predict(x_test2)
print('Mean squared error: Lin: %.2f'
      % mean_squared_error(y_test2, y_predLin))
print(y_predLin)
print("Number of mislabeled points out of a total %d points : Linear kernel :  %d"
      % (x_test2.shape[0], (y_test2 != np.round(y_predLin)).sum()))

#accuracy_score not working, calculate manually
acc_lin = (y_test2 == np.round(y_predLin)).sum()/x_test2.shape[0]

print('Accuracy Score: Linear kernel : '
 #     % accuracy_score(y_test2, y_predLin))
     + str(acc_lin))

Mean squared error: Lin: 69862030.02
[ 5635.73750534 14549.95825424 12421.78843345 ...  -231.78729973
 11553.45499458 12923.715279  ]
Number of mislabeled points out of a total 6302 points : Linear kernel :  6297
Accuracy Score: Linear kernel : 0.0007933989209774674


#### Poly

In [ ]:
svr_poly = SVR(kernel = 'poly')
svr_poly.fit(x_train2, y_train2)
y_predPoly = svr_poly.predict(x_test2)
print('Mean squared error: Poly: %.2f'
      % mean_squared_error(y_test2, y_predPoly))

print("Number of mislabeled points out of a total %d points : Polynomial kernel :  %d"
      % (x_test2.shape[0], (y_test2 != np.round(y_predPoly)).sum()))
print('Coefficient of determination: Polynomial:  %.2f'
      % r2_score(y_test2, y_predPoly))

acc_poly = (y_test2 == np.round(y_predPoly)).sum()/x_test2.shape[0]
print('Accuracy Score: Polynomial kernel :'
      + str(acc_poly))

#### RBF

In [ ]:
svr_rbf = SVR(kernel = 'rbf')
svr_rbf.fit(x_train2, y_train2)
y_predRbf = svr_rbf.predict(x_test2)

print('Mean squared error: RBF: %.2f'
      % mean_squared_error(y_test2, y_predRbf))

print(y_predRbf)

#np.round makes it correct, i think since SVR is a regression, it tries to extimate exactly, which is'nt an integer

print("Number of mislabeled points out of a total %d points : RBF kernel :  %d"
      % (x_test2.shape[0], (y_test2 != np.round(y_predRbf)).sum()))


print('Coefficient of determination: RBF:  %.2f'
      % r2_score(y_test2, y_predRbf))

acc_rbf = (y_test2 == np.round(y_predRbf)).sum()/x_test2.shape[0]

print('Accuracy Score: RBF kernel :'
      + str(acc_rbf))



[-0.90016884 -0.90020146 -0.90046147 ... -0.89962424 -0.90009629
 -0.899679  ]
 
Number of mislabeled points out of a total 6302 points : Linear kernel :  6302

Number of mislabeled points out of a total 6302 points : Polynomial kernel :  5652

Number of mislabeled points out of a total 6302 points : RBF kernel :  1959

Coefficient of determination: Linear:  -23340845.91

Coefficient of determination: Polynomial:  -1.31

Coefficient of determination: RBF:  -0.12

Accuracy Score: Linear kernel : 0.0

Accuracy Score: Polynomial kernel :0.10314185972707077

Accuracy Score: RBF kernel :0.6891463027610283
